### PR TITLE
qwen4exp: backport upstream correctness fixes (#27941)

### PR DESCRIPTION
**Pull request by amangupta-tether**
## Overview

Backport [ggml-org/llama.cpp PR #27941](https://github.com/ggml-org/llama.cpp/pull/27941), commit `36b10154383b60eb15baac2c7a40d2a5f784faa7`, onto `temp-10549`.

The upstream fix groups image cells by rank when M-RoPE repeats temporal positions, uses actual block-start spatial coordinates, and separates QSA blocks by sequence set. It also restores indexer keys during sequence copies, preserves cached spatial metadata on state restore, validates malformed model metadata, and avoids the pooled-normalization CUDA grid limit at long context.

Addresses the [multimodal QSA issue reported in PR #218](https://github.com/tetherto/qvac-fabric-llm.cpp/pull/218#issuecomment-5451100203).

## Additional information

Two cherry-pick conflicts were adapted to this branch:

- `src/llama-arch.cpp`: retain the existing Qwen3Next tensor-split restriction and add upstream's Qwen4Exp restriction. Qwen4Exp `-sm tensor` is disabled by this backport.
- `tests/test-save-load-state.cpp`: add upstream's byte-for-byte state round-trip test to the existing harness as test 6, without importing unrelated harness changes.

Validation on Linux arm64 CPU:

- `test-kv-cache` and `test-kv-cells`: passed.
- `test-llama-archs --arch qwen4exp --seed 1`: passed inference and model round-trip checks.
- All six save/load tests passed with a generated Qwen4Exp model in unified (`-np 1`) and separate-stream (`-np 2`) configurations.
- An ad hoc image-block check passed for rank grouping over fragmented physical cells, block-start M-RoPE coordinates, incomplete tails, and causal visibility.

GPU validation, full CI, and real-model performance/perplexity checks have not been run. The upstream image rank path remains limited to a single sequence in the cache stream; this backport does not expand that coverage.

## Requirements

- [x] I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md).
- AI usage disclosure: YES. The upstream commit contains its original AI attribution. Codex resolved the two backport conflicts, ran the reported checks, and updated this PR at the contributor's request. Human review and ownership remain required.

### CODE REVIEW

File: src/llama-arch.cpp [hunk 1]
```diff
@@ -1117,6 +1117,7 @@ bool llm_arch_supports_sm_tensor(const llm_arch & arch) {
         case LLM_ARCH_KIMI_K3:
         case LLM_ARCH_QWEN3TTS:
         case LLM_ARCH_QWEN3NEXT:
+        case LLM_ARCH_QWEN4EXP:   // TODO: fix test-llama-archs
```
Comment (gianni-cor) [inline src/llama-arch.cpp:1120]: is this added again by mistake?

Comment (amangupta-tether) [reply src/llama-arch.cpp:1120]: No was added in master recently because of `test-llama-archs` failing. Let me look at fixing the test

File: src/llama-memory-hybrid-idx.cpp [hunk 2]
```diff
@@ -261,6 +266,324 @@ llama_kv_cache * llama_memory_hybrid_idx::get_mem_idx() const {
...
+                if (idx%r == 0) {
+                    grp_slot0[g] = (int32_t) j;
+                }
+            }
+        };
+
+        group_cells();
+
+        // mrope repeats one position across an image, so rank cells instead of using the position
+        if (dup && ubatch->is_pos_2d() && one_seq) {
```
Comment (gianni-cor) [inline src/llama-memory-hybrid-idx.cpp:417]: QSA groups every four cache cells into one block. Image tokens can share the same temporal M-RoPE position `t`, so the code must sort them by `(t, y, x)` and assign unique ranks before grouping.

That ranking currently runs only when the cache contains one sequence:

```cpp
if (dup && ubatch->is_pos_2d() && one_seq)
```

With unified KV and multiple sequences, `one_seq` becomes false. Ranking is skipped, repeated positions all occupy the same block slot, and no complete four-cell blocks are created. The cells fall into one unpooled fallback block and may be arbitrarily dropped by top-k selection.

The proper fix is sequence-local `(t, y, x)` ranking. A safe alternative is to disable QSA for unified multi-sequence caches or use `--no-kv-unified`.

The same condition is still present on upstream llama.cpp `master`.

File: src/llama-arch.cpp (hunk 1 above)
Comment (amangupta-tether) [reply src/llama-arch.cpp]: Intentional. This was explicitly requested in the upstream review to disable `-sm tensor` until its QWEN4EXP failure is resolved: https://github.com/ggml-org/llama.cpp/pull/27941#discussion_r3901941773. Later upstream branches still carry this exclusion, so this backport keeps it unchanged.

File: src/llama-memory-hybrid-idx.cpp (hunk 2 above)
Comment (amangupta-tether) [reply src/llama-memory-hybrid-idx.cpp:417]: Fixed in `118e5d7ac` using the safe fallback suggested here: QSA now stays enabled for a single sequence or per-sequence KV streams, while unified KV contexts configured for multiple sequences use dense attention. Added `test-qsa-unified-multiseq`, which verifies that no QSA top-k node is built for a two-sequence M-RoPE image batch and that its logits match an explicitly dense model.

File: src/llama-arch.cpp (hunk 1 above)
Comment (amangupta-tether) [reply src/llama-arch.cpp]: Fixed in `3caf2d585`.

The QWEN4EXP tensor-parallel math was already correct; the failure came from the Meta validation backend. Its host-view workaround only accepted views whose source op was `NONE`, while QWEN4EXP builds `model.input_embed (GET_ROWS) -> RESHAPE` in a host buffer. The scheduler can carry that no-op reshape into the Meta split, where the old code tried to map it as a Meta tensor and aborted.

I generalized the existing pass-through to all host-backed views and removed QWEN4EXP from the `-sm tensor` denylist.

Validation:
- focused CUDA / CPU / Meta QWEN4EXP test: pass (`8.65e-08` NMSE)
- temporary 2-worker tensor split on the single available GPU: pass (`8.65e-08`, internal AllReduce exercised)
- full CUDA / CPU / Meta architecture suite: pass
